core/include/utils/algorithm.hpp

```diff
@@ -18,35 +27,41 @@ namespace traccc {
  * to be able to transform the input type into the output type in one way or
  * another.
  *
- * @param I The input type of the algorithm
+ * @param A The input types of the algorithm
  * @param O The output type of the algorithm
  */
-template <typename I, typename O>
-class algorithm {
+template <typename R, typename... A>
+class algorithm<R(A...)> {
     public:
-    using input_type = I;
-    using output_type = O;
-
-    /**
-     * @brief Execute the algorithm, returning by value.
-     *
-     * Turn the object into a callable, accepting only an input and
-     * returning the output through value return.
-     *
-     * @param[in] i The input value to the algorithm
-     * @return An instance of the specified output type
-     */
-    virtual output_type operator()(const input_type& i) const = 0;
-
-    /**
-     * @brief Execute the algorithm, returning by input reference.
-     *
-     * Turn the object into a callable, accepting an input as well as an
-     * output parameter to write to.
-     *
-     * @param[in] i The input value to the algorithm
-     * @param[out] o The output of the algorithm
-     */
-    virtual void operator()(const input_type& i, output_type& o) const = 0;
+    using output_type = R;
+
+    using function_type = R(A...);
+
+    static_assert(
+        std::conjunction<rvalue_or_const_lvalue<A>...>::value,
+        "All arguments must be either affine types (rvalue references), or "
+        "immutable constant types (const lvalue references).");
+
+    virtual output_type operator()(A... args) const = 0;
 };
+
+template <typename A, typename B, typename C, typename... R>
+auto compose(std::function<std::remove_reference_t<B>(A)> f,
+             std::function<C(B)> g, R... rs) {
+    if constexpr (sizeof...(R) > 0) {
+        auto h = compose(g, rs...);
+        return [f, h](A&& i) { return h(f(std::forward<A>(i))); };
+    } else {
+        return [=](A&& i) { return g(f(std::forward<A>(i))); };
+    }
+}
+
+template <typename A, typename B, typename C>
+std::function<B(A&&)> side_effect(std::function<B(A)> f,
+                                  std::function<void(const C&)> s) {
+    return [=](A&& i) -> B {
+        s(i);
+        return f(std::forward<A>(i));
+    };
+}
```

Comment:
Do these two functions need to be part of `algorithm.hpp`? From what I see, they are only used in testing the code, no? Shouldn't they be defined in the test code, in a "hidden" way instead?

Reply:
Hmm, I planned to make quite extensive use of these higher-order tools to write tests, for example. I think they will be useful to keep around.